[Refactor] 일기 사진 수 변경 및 매핑 수정

yoonsu417 opened the pull request:
## #️⃣연관된 이슈
> #32 

## 📝작업 내용
> 업로드 시 사진 한 장으로 수정
> setDiaries()로 photo <-> diaries 양방향 매핑
> CommandService, CommandServiceImpl 로 클래스 명 통일

## 🔎코드 설명
> List<MultipartFile>을 MultipartFile로 수정

## 💬고민사항 및 리뷰 요구사항 (Optional)
>  
    public void setDiaries(Diaries diaries) {
        if (this.diaries != null && diaries.getDiaryPhotosList() != null) {
            this.diaries.getDiaryPhotosList().remove(this);
        }

        this.diaries = diaries;

        if (diaries != null && diaries.getDiaryPhotosList() != null) {
            diaries.getDiaryPhotosList().add(this);
        }
    }


    이 부분에서 조건문에 && diaries.getDiaryPhotosList() != null 이 부분을 추가하지 않으면 
    setDiaries를 호출할 때 Cannot invoke "java.util.List.remove(Object)" because the return value of "TtattaBackend.ttatta.domain.Diaries.getDiaryPhotosList()" is null 이런 오류 발생
    
    그래서 코드를 저렇게 바꿨는데, 이렇게 바꿔도 괜찮은지... 고민입니다.

## 비고 (Optional)
> x

yoonsu417 committed:
refactor: #32 일기 작성 API 수정

업로드 시 사진 여러 장 -> 한 장으로 수정

yoonsu417 committed:
Merge remote-tracking branch 'origin/develop' into refactor/#32-diaries-save

# Conflicts:
#	src/main/java/TtattaBackend/ttatta/service/DiaryService/DiaryPhotoService.java
#	src/main/java/TtattaBackend/ttatta/service/DiaryService/DiaryPhotoServiceImpl.java

yoonsu417 committed:
refactor: #32 일기 사진 한장 변경 및 photo 매핑

save: 사진 한 장 저장 변경
setDiaries 로 양방향 매핑

yoonsu417 committed:
refactor: #32 일기 관련 클래스명 수정

CommandService, CommandServiceImpl로 변경

Jinho622 commented:
DiaryConverter의 toDiaries메서드에서 builder를 통해 diaries객체가 생성될 때 diaryPhotosList 객체 변수가 new ArrayList로 초기화되지 않고 있는게 근본적인 문제인 것 같아요. 그래서 '"TtattaBackend.ttatta.domain.Diaries.getDiaryPhotosList()" is null'이란 오류가 뜨고 있는 것 같아요. 그리고 diaryPhoto객체의 diaries객체 변수에 중복해서 값을 넣고 있는데 이는 에러를 발생시키지는 근본적인 원인은 아니지만 논리적으로 봤을 때 수정되어야 될 것 같아요. 어쨌든 다음방법으로 오류 해결해서 공유드립니다! 

## DiaryPhotos 엔티티에서 매핑을 위한 diaries객체에 값이 중복으로 들어가는 것 수정
DiaryCommandServiceImpl: DiaryPhotos 엔티티의 setDiaries함수에서 diaries라는 객체변수에 값을 넣어주고 있어서 DiaryConverter의 toDiaryPhoto메서드에서 builder로 DiaryPhotos객체를 생성할 때 아래 사진과 같이 객체변수로 diaries를 넣어줄 필요가 없어요.
![image](https://github.com/user-attachments/assets/1f54e559-f257-4ade-af63-06e93624bf78)

DiaryPhotos 엔티티의 diaries에 값이 들어간 상태로 setDiaries가 호출되고 있어서 불필요하게 setDiaries메서드의 if문(아래 사진의 if문)이 돌아가고 있어요.
![image](https://github.com/user-attachments/assets/153390f7-3653-4b20-ab57-6c767a989f91)
그래서 this.diaries.getDiaryPhotosList().remove(this); 가 실행이 되는데 diaries엔티티의 diaryPhotosList가 초기화되지 않은 근본적인 문제 때문에 윤수님이 올려주신 에러가 발생하고 있어요.

![image](https://github.com/user-attachments/assets/cd08f4e5-a7bc-4385-ba23-ea969c363faa)
때문에 위 사진처럼 DiaryConverter의 toDiaryPhoto에서 .diaries(diaries)부분을 주석처리를 하면 if문에 걸리지 않아 'Cannot invoke "java.util.List.remove(Object)" because the return value of "TtattaBackend.ttatta.domain.Diaries.getDiaryPhotosList()" is null'라는 오류는 해결돼요.
하지만 diaries엔티티의 diaryPhotosList가 초기화되지 않은 근본적인 문제는 아직 해결되지 않아 이를 해결해야 돼요.


## diaryPhotosList가 초기화되지 않는 근본적인 문제 해결 방법
근데  diaryPhotosList가 초기화되지 않는 근본적인 문제를 해결하지 않아 "Cannot invoke \"java.util.List.add(Object)\" because the return value of \"TtattaBackend.ttatta.domain.Diaries.getDiaryPhotosList()\" is null"라는 다른 에러가 발생하는데요..(이건 DiaryPhotos 엔티티의 setDiaries함수에서 아래 사진의 if문에서 발생하고 있어요..!)
![image](https://github.com/user-attachments/assets/898a677e-87b1-44f9-9380-4a4252f174fc)

이는 DiaryConverter의 toDiaries메서드에서 diaries객체를 만들 때 diaryPhotosList가 초기화되지 않아 문제가 발생하고 있어요.(builder로 객체를 생성할때는 직접 명시해주지 않는다면 객체 변수가 클래스에 있는대로 초기화 되지 않는다고 합니다..!)
그래서 DiaryConverter의 toDiaries메서드에서 builder로 diaries객체를 만들 때 아래 사진 처럼 diaryPhotosList를 직접 초기화시켜주면 해당 에러는 해결됩니다!
![image](https://github.com/user-attachments/assets/cd1616db-7a7c-414a-99cb-47792916dd3e)

수정된 코드는 제가 이 브랜치에 push해둘게요. 수정된 코드 확인 부탁드립니다.

Jinho622 commented:
그리고 diaryPhotos테이블과 uuid테이블을 매핑(양방향 매핑)해서 diaries의 값이 삭제될 때 연관된 diaryPhotos뿐만아니라 연관된 uuid도 삭제되게끔(cascade로 삭제가 되도록..!) 설정해주셔야 될 것 같아요.
사진 여러장일때 필요한 기능이고 필수로 있어야하는 기능은 아니어서 다른 것 먼저 하시고 시간 나시면 부탁드려요!

Jinho622 committed:
fix: #32 DiaryConverter의 toDiaries메서드에서 Diaries엔티티를 builder로 생성할 때 diaryPhotosList가 초기화되지 않는 문제 해결

해결 방법: builder에서 diaryPhotosList를 new ArrayList<>()로 직접 초기화함
위의 문제 외에 diaryPhoto객체의 diaries객체 변수에 중복해서 값을 넣고 있는 논리적 오류 해결
해결 방법: DiaryPhotos엔티티의 setDiaries에서 매핑을 위해 diaries변수에 값을 넣어주고 있으니 DiaryConverter의 toDiaryPhoto메서드의 builder에서는 diaries를 넣지 않아야 함

yoonsu417 commented:
@Jinho622 코드 수정해주신 거 확인했습니다 !!! 감사합니다 :)  uuid는 사진 여러장 구현할 때 추가로 수정하겠습니다.

Jinho622 committed:
fix: #32 Diaries domain의 @OneToMany설정에서 'fetch = FetchType.EAGER'는 불필요하여 제거

Jinho622 committed:
fix: #32 Diaries 도메인의 setDiaryPhotosList메서드 또한 불필요하여 제거

Jinho622 commented:
넵 감사합니다. approve했으니 merge부탁드려요!
